[Feature] Allow actions to be shared via public link: Milestone 1

calherries opened the pull request:
This PR targets milestone 1 of "Allow actions to be shared via public link".

It includes the following changes:

***Milestone 1 (Product to review when this is done)***
- [x] Add **Make public** toggle to the action settings modal
  - [x] FE: https://github.com/metabase/metabase/pull/27809
  - [x] BE: #27714
- [x] Create a desktop public action page
  - [x] BE: #27723 
  - [x] BE: #27717
    - [x] Make sure the endpoint doesn't execute if actions are disabled for the database
  - [x] FE: #27747
- [x] Show thank you, error message after action form submission
  - [x] FE: #27747

And also these originally planned for Milestone 2:
- [x] Show labels for publicly accessible actions on the model detail page
  - [x] FE: https://github.com/metabase/metabase/pull/28039
  - [x] BE: https://github.com/metabase/metabase/issues/27718

Epic: https://github.com/metabase/metabase/issues/27626

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/27777)
<!-- Reviewable:end -->

calherries committed:
Add public_uuid and made_public_by_id to actions table & endpoints to enable/disable the public sharing on individual actions (#27721)

* Add migration for public_uuid (indexed) and made_public_by_id

* Add UUIDString schema

* Add endpoints for enabling/disabling sharing of actions

* Test that the new fields are returned with `GET /api/action/:id`

* Remove validCheckSum on migration

* Remove trailing whitespace

* Fix DELETE test

* Fix tests

* Please migration linter

* Update the default public_uuid every test run

* Please migration linter

* Add FK onDelete cascade

* Replace ü

Commit:
Add GET endpoint and post-select for action

Commit:
Revert "Add GET endpoint and post-select for action"

This reverts commit 8cc8b57d6034146dd726b54bc4199830ec1fda21.

calherries committed:
Merge branch 'master' into public-actions-feature

calherries committed:
Fix merge

Commit:
Import metabase.actions.test-utils vars to metabase.test (#27767)

* Add import vars to metabase.test

* Adjust clj-kondo config

* Import more vars to metabase.test

* Remove unused vars from clj-kondo

* Import another var

* Replace actions.test-util with mt

* Fix merge

* Import another var

* Resolve cyclic deps

* Sort ns

* Fix quoted namespace

Commit:
Reorder migrations

calherries committed:
Update test for GET `api/action?model-id=<id>` endpoint to include public sharing keys (#27802)

calherries committed:
Add GET: /api/public/action/:uuid endpoint (#27781)

* Add test

* Remove non-public columns and check for 404

* Fix docstring

* Rename for clarity

* Fix missing ns

* Remove unneccessary keys from action

* Update test

* Remove unused refer

* Reorder migrations

* Use mt

* Add require and refactor

* Use mt

calherries committed:
Add endpoint for executing a public action (#27793)

* Add endpoint and test

* Add more tests and TODOs

* Use mt

* Reorder migrations

* Remove unused require

* Rate limit actions at 1 per second

* Fix the tests for the throttle

* Refactor tests

* Fix test

Commit:
Fix docstring

Commit:
Merge branch 'master' into public-actions-feature

Commit:
Add test for failed execution if actions are disabled

Commit:
Use crowberto in tests

Commit:
Fix using crowberto in tests

Commit:
Fix cyclic load dependency

WiNloSt committed:
Refactor ActionCreator (#27832)

* Refactor make ActionCreator more sane

* Render sidebar conditionally with JS rather than hiding it in CSS

kulyk commented:
@calherries should `GET /api/public/action/:uuid` return 404 instead of 400 if there's no action with a given UUID?

calherries commented:
@kulyk Yes, but that's a separate BE task and solving it has implications for dashboards and cards. This is an issue with all `GET /api/public/***/:uuid` endpoints.

There's a comment Cam left a while back here. https://github.com/metabase/metabase/blob/84b55aeeb2b514ed96744ad0dfe918fe267e3d05/test/metabase/api/public_test.clj#L114-L115

I'm choosing to punt on it.

Commit:
Merge branch 'master' into public-actions-feature

WiNloSt committed:
Merge branch 'master' into public-actions-feature

WiNloSt committed:
Make action public (#27809)

* Move action creator action buttons to the header following design

* Remove double border which make it looks thicker

* Draft toggle action public

* Add confirmation when disabling public link similiar to questions and dashboards

* Add an action public UUID input and copy button

* Show action settings based on user permission

* Add public action toggle tests

* Remove unused import 🤦‍

* Attempt to fix flaky CI unit tests

The problem seems to be because of how long it takes for the response
to be received even on unit tests, it took longer than 1 second which is
the default timeout for `waitFor`.

* See if not using `userEvent` could make the involving nock faster

* Improve test speed to reduce flakiness

kulyk committed:
Add public action page (#27747)

* Add basic public action page

* Add form submit logic placeholder

* Use "Submit" as default action form's button label

* Add "big" variant to `EmbedFrame's` footer

* Use new footer variant for action page

* Break down the page, add document title

* Handle long forms better

* Add `PublicWritebackAction` type

* Use public action GET endpoint

* Add endpoint to execute public actions

* Use action execution endpoint

* Add tests

* Handle actions without parameters

* Rename variant prop

* Replace `waitFor` with `findBy`

* Define `FooterVariant` type

* Fix router setup in `renderWithProviders`

WiNloSt committed:
Show which action is publicly accessible on model detail page (#28039)

kulyk committed:
Bring e2e tests back

Manually bringing back tests added in: https://github.com/metabase/metabase/pull/28056

tsmacdonald commented on src/metabase/api/public.clj:
```diff
@@ -531,6 +553,44 @@
+                           (catch ExceptionInfo e
+                             (get-in (ex-data e) [:errors :action-uuid])))
+        throttle-time (when throttle-message
+                        (second (re-find #"You must wait ([0-9]+) seconds" throttle-message)))]
```
This seems correct to me, but it feels like [`calculate-delay`](https://github.com/metabase/throttle/blob/master/src/throttle/core.clj#L121) should be made public and used directly? @camsaul any thoughts?

calherries replied:
Or we could make `[make-throttling-ex](make-throttling-ex)` return a map with the `delay-ms` and the message in two values.

calherries committed:
Update typo

Co-authored-by: Tim Macdonald <tim@metabase.com>

tsmacdonald commented on src/metabase/api/public.clj:
```diff
@@ -531,6 +553,44 @@
+        (binding [api/*current-user-permissions-set* (delay #{"/"})]
+            ;; Undo middleware string->keyword coercion
+          (let [action (api/check-404 (first (action/actions-with-implicit-params nil :public_uuid uuid)))]
+            (actions.execution/execute-action! action (update-keys parameters name))))))))
```
Weak preference for pulling this out into an `execute-action-by-uuid` helper fn to avoid the `do` and create some space with the throttling, as in `api.session/forgot-password-impl`

tsmacdonald commented on src/metabase/api/public.clj:
```diff
@@ -531,6 +553,44 @@
+               :body   throttle-message}
+        throttle-time (assoc :headers {"Retry-After" throttle-time}))
+      (do
+        ;; TODO: check-actions-enabled for the database
```
Feels like this should happen in this PR? If not, is the work ticketed?

calherries replied:
Ah, this is implemented. The comment can be deleted.

tsmacdonald commented on src/metabase/api/public.clj:
```diff
@@ -531,6 +553,44 @@
+   `extra_parameters` should be the extra, user entered parameter values."
+  [uuid :as {{:keys [parameters], :as _body} :body}]
+  {uuid        su/UUIDString
+   parameters  (s/maybe {s/Keyword s/Any})}
```
Why not make this a Malli schema too?

tsmacdonald replied:
Sorry, that sounded passive-aggressive. Since you did the work to define `UUIDString` twice I assume there's a good reason and so I'm curious

calherries replied:
I was probably just being lazy. I've updated it to use Malli now, thanks

tsmacdonald reviewed (approved):
A few comments but nothing blocking; change (or not) as you see fit

qnkhuat commented on src/metabase/api/action.clj:
```diff
@@ -139,4 +143,28 @@
+  Action has already been shared, it will return the existing public link rather than creating a new one.) Public
+  sharing must be enabled."
+  [id]
+  (api/check-superuser)
```
to confirm: we still working on the exact permission model for public actions, requiring admin now just because it's the 1st milestone right?

qnkhuat commented on src/metabase/api/action.clj:
```diff
@@ -139,4 +143,28 @@
+  sharing must be enabled."
+  [id]
+  (api/check-superuser)
+  (validation/check-has-application-permission :setting)
```
why setting permissions are required here? isn't admins always have this permission?

calherries replied:
Correct, fixed

qnkhuat commented on src/metabase/api/action.clj:
```diff
@@ -139,4 +143,28 @@
+  "Generate publicly-accessible links for this Action. Returns UUID to be used in public links. (If this
+  Action has already been shared, it will return the existing public link rather than creating a new one.) Public
+  sharing must be enabled."
+  [id]
```
nit: could we have schema for this? I believe  it'll be ```{id ms/IntGreaterThanZero}```

calherries replied:
Done

qnkhuat commented on src/metabase/api/action.clj:
```diff
@@ -139,4 +143,28 @@
+
+(api/defendpoint DELETE "/:id/public_link"
+  "Delete the publicly-accessible link to this Dashboard."
+  [id]
```
should we require admin for this?

calherries replied:
See this comment
https://github.com/metabase/metabase/blob/f02a110b5588c2218c81b86cf44ea7ae00cccd25/src/metabase/api/action.clj#L166

qnkhuat commented on src/metabase/api/public.clj:
```diff
@@ -304,6 +306,26 @@
+  "Fetch a publicly-accessible Action. Does not require auth credentials. Public sharing must be enabled."
+  [uuid]
+  {uuid ms/UUIDString}
+  (validation/check-public-sharing-enabled)
```
Should we have `(validation/check-has-application-permission :setting)` here? i

qnkhuat commented on src/metabase/api/public.clj:
```diff
@@ -531,6 +553,44 @@
+        ;; Run this query with full superuser perms. We don't want the various perms checks
+        ;; failing because there are no current user perms; if this Dashcard is public
+        ;; you're by definition allowed to run it without a perms check anyway
+        (binding [api/*current-user-permissions-set* (delay #{"/"})]
```
shouldn't this be an `atom` instead of delay?

calherries replied:
I'm not sure to be honest. Both `delay` and `atom` are used throughout the codebase. But I can't see any reason to use `atom` over `delay`, since it doesn't appear to be swapped / reset anywhere.

qnkhuat replied:
I think we only use `delay` in tests. 

In apis, we use `atom`, there are good amount of this pattern in `api/public`

tsmacdonald replied:
My understanding is that `delay` is the better choice when possible since it communicates that you're not mutating it. (And of course the `delay` is necessary at all since the value gets derefed)

calherries replied:
I'm not sure why the value needs to be a `delay` or `atom` at all, which is why the choice is confusing. Maybe it's for historical reasons. Could it just be a dynamic var?

qnkhuat replied:
we used `delay` as an optimization for APIs don't need to query permissions.

The same for `curernt-user, user-locale` : https://github.com/metabase/metabase/blob/d5c768af1263fc98c84ff17877b7d50972a8e26d/src/metabase/server/middleware/session.clj#L312

qnkhuat replied:
@tsmacdonald good reasoning. Yea I agree it's better to use `delay` here now.

qnkhuat reviewed (approved):
I think BE looks great. The tests are very tight!

I have a couple of comments/questions about the permissions check but shouldn't be blockers.

Commit:
Remove TODO and use malli for defendpoint

Commit:
Add malli schema for endpoints

Commit:
Update permissions checks for POST and DELETE

Commit:
Move comment

kulyk committed:
Fix `ActionCreator` tests
